### PR TITLE
Add HttpClient

### PR DESCRIPTION
**Pull request by foziljonovs**

### CODE REVIEW

**Comment by foziljonovs** [comment]
Man davom  ettirishim uchun qiziqishlar frontEnd tugashini kutyapman

File: IKwizApiHttpClient.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+public interface IKwizApiHttpClient
+{
+    Task<UserInterest[]> GetUserInterestsAsync();
+    Task<bool> IsUserInterestsSubmittedAsync();
+    Task SubmitUserInterestAsync(UserInterest interests);
```
Comment (wahidustoz) [inline IKwizApiHttpClient.cs]:
```suggestion
    Task SubmitUserInterestsAsync(UserInterest interests);
```

File: IKwizApiHttpClient.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+public interface IKwizApiHttpClient
+{
+    Task<UserInterest[]> GetUserInterestsAsync();
```
Comment (wahidustoz) [inline IKwizApiHttpClient.cs]:
```suggestion
    Task<IEnumerable<UserInterest>> GetUserInterestsAsync();
```

File: KwizApiHttpClient.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+using System.Linq;
+using System.Net.Http.Json;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Kwiz;
+
+public class KwizApiHttpClient : IKwizApiHttpClient
+{
+    private readonly HttpClient client;
+    private UserInterest[]? userInterest;
```
Comment (wahidustoz) [inline KwizApiHttpClient.cs]:
```suggestion
```

File: KwizApiHttpClient.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    public KwizApiHttpClient(HttpClient client)
+    {
+        this.client = client;
+    }
```
Comment (wahidustoz) [inline KwizApiHttpClient.cs]:
```suggestion
    public KwizApiHttpClient(IHttpClientFactory httpClientFactory)
    {
        this.client = httpClientFactory.CreateClient("KwizApi");
    }
```

File: IKwizApiHttpClient.cs
Comment (wahidustoz) [inline IKwizApiHttpClient.cs:1]: Quyidagi methodlar implement bo'lishi kerak:

- `ValueTask<IEnumerable<UsersInterest>> GetUserInterestsAsync()` 
    - this method calls GetUserInterestsOrDefaultAsync(), if null returned throws UserInterestsNotSubmittedException()
- `ValueTask<IEnumerable<UsersInterest>> GetUserInterestsOrDefaultAsync()`
    - this method returns null if API returns 204 NO Content
    - this method uses **Polly** to retry http request incase request fails
        - only retry http timeout exception

File: Program.cs
Comment (wahidustoz) [inline Program.cs:1]:
```suggestion
builder.Services.AddScoped<CustomAuthorizationMessageHandler>();
builder.Services.AddHttpClient<IKwizApiHttpClient, KwizApiHttpClient>(client => client.BaseAddress = new Uri(builder.Configuration["KwizApiBaseUrl"]))
    .AddHttpMessageHandler<CustomAuthorizationMessageHandler>();
```

File: UserInterest.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+public class UserInterest
+{
+    public Guid UserId { get; set; }
+    public string[] Interests { get; set; }
```
Comment (wahidustoz) [inline UserInterest.cs]:
```suggestion
    public IEnumerable<string> Interests { get; set; }
```

**Review by wahidustoz — APPROVED** [review]

**Pull request merged**